File: src/ocr_merge_pages.py
import sys
import re
import argparse
from pathlib import Path

def merge_ocr_pages(file_path: Path):
    """
    OCRで出力されたMarkdownファイルからページ区切りマーカーを処理（結合・削除）し、
    整形された _clean.md ファイルとして保存する。
    """
    if not file_path.exists():
        print(f"[ERROR] File not found: {file_path}")
        return

    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        # 1. 新しいページ区切り形式の処理
        # パターン: =-- End Printed Page ... --= (改行/空白) =-- Begin Page ... --=
        # どちらかに (Continuation) があれば結合する
        
        # 境界マーカーとその前後の空白（改行含む）をマッチさせる
        boundary_pattern = r'\s*(=-- End Printed Page [^\n]*? --=)\s*(=-- Begin Page [^\n]*? --=)\s*'
        
        def replace_boundary(match):
            end_marker = match.group(1)
            begin_marker = match.group(2)
            
            if "(Continuation)" in end_marker or "(Continuation)" in begin_marker:
                # Continuationがある場合は、マーカーとその前後の空白を全て削除（段落結合）
                return ""
            else:
                # Continuationがない場合は、マーカーを削除し、空行(段落区切り)を入れる
                return "\n\n"

        content = re.sub(boundary_pattern, replace_boundary, content, flags=re.DOTALL)

        # 2. 残った単独のマーカー（ファイルの先頭や末尾など）を削除
        content = re.sub(r'=-- Begin Page [^\n]*? --=\s*', '', content)
        content = re.sub(r'\s*=-- End Printed Page [^\n]*? --=', '', content)

        # 3. 旧形式のマーカー処理（念のため残しておく）
        cont_pattern = r'\s*^=-- Page .*?\(Continuation\).*?--=\s*'
        content = re.sub(cont_pattern, '', content, flags=re.MULTILINE)
        normal_pattern = r'^=-- Page .*?--=\s*\n?'
        content = re.sub(normal_pattern, '', content, flags=re.MULTILINE)
        
        new_content = content
        
        # 3つ以上の連続する改行を2つ（1つの空行）に置換して整える
        new_content = re.sub(r'\n{3,}', '\n\n', new_content)
        
        # 出力ファイル名 (例: file_paged.md -> file.md)
        if file_path.name.endswith("_paged.md"):
            output_path = file_path.with_name(file_path.name.replace("_paged.md", ".md"))
        else:
            output_path = file_path.with_name(file_path.stem + "_merged" + file_path.suffix)
        
        with open(output_path, 'w', encoding='utf-8') as f:
            f.write(new_content)
            
        print(f"[SUCCESS] Created: {output_path}")

    except Exception as e:
        print(f"[ERROR] Failed to process {file_path}: {e}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="OCRで出力されたMarkdownのページ区切りを処理し、テキストを結合・整形します。")
    parser.add_argument("input_path", type=str, help="対象のMarkdownファイルまたはディレクトリ")
    
    args = parser.parse_args()
    input_path = Path(args.input_path)
    
    if input_path.is_dir():
        md_files = list(input_path.glob("*_paged.md"))
        if not md_files:
            # _paged.md が見つからない場合は全ての .md を対象にする
            md_files = list(input_path.glob("*.md"))
            
        print(f"[INFO] Found {len(md_files)} Markdown files in {input_path}")
        for md_file in md_files:
            # 既に処理済みのファイルや、意図しないファイルを除外
            if md_file.name.endswith("_merged.md") or md_file.name.endswith("_clean.md"):
                continue
            merge_ocr_pages(md_file)
    else:
        merge_ocr_pages(input_path)

File: src/ocr_merge_pages.js
/**
 * OCRで出力されたMarkdownファイルからページ区切りマーカーを処理（結合・削除）し、
 * 段落を整形して保存するプログラム。
 * 
 * 使い方:
 *   node src/ocr_merge_pages.js <Markdownファイルパス または ディレクトリパス>
 */
const fs = require('fs');
const path = require('path');

/**
 * OCRで出力されたMarkdownファイルからページ区切りマーカーを処理（結合・削除）し、
 * 整形されたファイルとして保存する。
 */
function mergeOcrPages(filePath) {
    if (!fs.existsSync(filePath)) {
        console.error(`[エラー] ファイルが見つかりません: ${filePath}`);
        return;
    }

    try {
        let content = fs.readFileSync(filePath, 'utf-8');

        // ページ番号の整合性チェック
        const beginPagePattern = /### -- Begin Page (\d+).*? --/g;
        const endPagePattern = /### -- End .*? --/g;

        // 物理ページと印字ページの対応関係を抽出
        const pageInfo = [];
        const pageBlocks = content.split(/### -- Begin Page \d+.*? --/);
        const beginMarkers = content.match(beginPagePattern) || [];
        
        // splitの結果、最初の要素は最初のBegin Pageより前の内容（通常は空）
        for (let i = 0; i < beginMarkers.length; i++) {
            const marker = beginMarkers[i];
            const pageContent = pageBlocks[i + 1] || "";
            
            const physicalMatch = marker.match(/### -- Begin Page (\d+)/);
            const physicalNum = physicalMatch ? parseInt(physicalMatch[1]) : null;
            
            let printedNum = null;
            const endMatch = pageContent.match(/### -- End .*? --/);
            if (endMatch) {
                const pm = endMatch[0].match(/\(Printed Page (\d+)\)/);
                if (pm) printedNum = parseInt(pm[1]);
            }
            
            if (physicalNum !== null) {
                pageInfo.push({ physical: physicalNum, printed: printedNum });
            }
        }

        const warnings = [];

        // Begin Page (物理連番) のチェック
        if (pageInfo.length > 0) {
            for (let i = 0; i < pageInfo.length; i++) {
                if (i > 0 && pageInfo[i].physical !== pageInfo[i - 1].physical + 1) {
                    warnings.push(`[警告] Begin Pageの連番に飛びがあります: ${pageInfo[i - 1].physical} -> ${pageInfo[i].physical}`);
                }
            }
            if (pageInfo[0].physical !== 1) {
                warnings.push(`[警告] Begin Pageが1から始まっていません (開始番号: ${pageInfo[0].physical})`);
            }
        }

        // 印字ページ（Printed Page）の整合性チェック
        const printedInfo = pageInfo.filter(p => p.printed !== null);
        if (printedInfo.length > 1) {
            for (let i = 1; i < printedInfo.length; i++) {
                const prev = printedInfo[i - 1];
                const curr = printedInfo[i];
                
                const physicalDiff = curr.physical - prev.physical;
                const printedDiff = curr.printed - prev.printed;

                if (printedDiff <= 0) {
                    warnings.push(`[警告] 印字ページ番号が逆転または重複しています: 物理${prev.physical}P(印字${prev.printed}P) -> 物理${curr.physical}P(印字${curr.printed}P)`);
                } else if (printedDiff !== physicalDiff) {
                    warnings.push(`[警告] 印字ページ番号の進みが物理ページと一致しません（ページ抜けの疑い）: 物理${prev.physical}P(印字${prev.printed}P) -> 物理${curr.physical}P(印字${curr.printed}P)`);
                }
            }
        }

        // 1. 新しいページ区切り形式の処理
        // パターン: ### -- End ... -- (改行/空白) ### -- Begin Page ... --
        const boundaryPattern = /\s*(### -- End [\s\S]*? --)\s*(### -- Begin Page [\s\S]*? --)\s*/g;
        
        content = content.replace(boundaryPattern, (match, endMarker, beginMarker) => {
            if (endMarker.includes("(Continuation)") || beginMarker.includes("(Continuation)")) {
                // Continuationがある場合は、マーカーとその前後の空白を全て削除（段落結合）
                return "";
            } else {
                // Continuationがない場合は、マーカーを削除し、空行(段落区切り)を入れる
                return "\n\n";
            }
        });

        // 2. 残った単独のマーカー（ファイルの先頭や末尾など）を削除
        content = content.replace(/### -- Begin Page [\s\S]*? --\s*/g, '');
        content = content.replace(/\s*### -- End [\s\S]*? --/g, '');

        // 3. 旧形式のマーカー処理
        content = content.replace(/\s*^=-- Page .*?\(Continuation\).*?--=\s*/gm, '');
        content = content.replace(/^=-- Page .*?--=\s*\n?/gm, '');
        
        // 3つ以上の連続する改行を2つ（1つの空行）に置換して整える
        content = content.replace(/\n{3,}/g, '\n\n');
        
        // 出力ファイル名
        let outputPath;
        const fileName = path.basename(filePath);
        const dirName = path.dirname(filePath);
        const ext = path.extname(filePath);
        const stem = path.basename(filePath, ext);

        if (fileName.endsWith("_paged.md")) {
            outputPath = path.join(dirName, fileName.replace("_paged.md", "_merged.md"));
        } else {
            outputPath = path.join(dirName, stem + "_merged" + ext);
        }
        
        fs.writeFileSync(outputPath, content, 'utf-8');
        console.log(`[成功] 作成されました: ${outputPath}`);

        // 警告のレポート
        if (warnings.length > 0) {
            console.log(`\n--- ${fileName} のページ整合性レポート ---`);
            warnings.forEach(w => console.warn(w));
            console.log("-----------------------------------------------\n");
        }

    } catch (err) {
        console.error(`[エラー] ${filePath} の処理に失敗しました: ${err}`);
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.log("-------------------------------------------------------");
        console.log(" Markdownファイル（_paged.md）またはフォルダをドロップしてください。");
        console.log(" 使い方: node ocr_merge_pages.js <input_path...>");
        console.log("-------------------------------------------------------");
        return;
    }

    for (const arg of args) {
        const inputPath = path.resolve(arg);
        if (!fs.existsSync(inputPath)) {
            console.error(`[エラー] パスが見つかりません: ${inputPath}`);
            continue;
        }
        
        if (fs.statSync(inputPath).isDirectory()) {
            let mdFiles = fs.readdirSync(inputPath)
                .filter(f => f.endsWith("_paged.md"))
                .map(f => path.join(inputPath, f));
                
            if (mdFiles.length === 0) {
                mdFiles = fs.readdirSync(inputPath)
                    .filter(f => f.endsWith(".md"))
                    .map(f => path.join(inputPath, f));
            }
                
            console.log(`[情報] ${inputPath} 内に ${mdFiles.length} 個の Markdown ファイルが見つかりました`);
            for (const mdFile of mdFiles) {
                if (mdFile.endsWith("_merged.md")) {
                    continue;
                }
                mergeOcrPages(mdFile);
            }
        } else {
            mergeOcrPages(inputPath);
        }
    }
    console.log("\nすべての処理が完了しました。");
}

if (require.main === module) {
    main();
}
